File: data_svc/patches/cdp_discover.js
/**
 * Intelligent CDP port discovery for the TradingView Desktop integration.
 *
 * Mirror of data_svc/cdp_discover.py — same algorithm and the same cache file
 * format so the Python data-svc and this Node MCP server converge on the same
 * port without coordination.
 *
 * Algorithm:
 *   1. If env CDP_PORT is set, use it verbatim (operator escape hatch).
 *   2. Probe the cached port (if a cache file exists). If host:port still
 *      hosts a TradingView instance, return it (fast path).
 *   3. Scan the configured port range. The first port whose /json/list
 *      contains a tradingview tab wins.
 *   4. Write the chosen endpoint to the cache file atomically.
 *
 * Env overrides:
 *   CDP_PORT             explicit port (skip discovery)
 *   CDP_HOST             host (defaults to "localhost")
 *   TV_MCP_CACHE_FILE    cache file path override
 *   TV_MCP_PORT_RANGE    port range "low-high" (default "9222-9230")
 */

import fs from 'fs/promises';
import fssync from 'fs';
import http from 'http';
import os from 'os';
import path from 'path';

const DEFAULT_HOST = 'localhost';
const DEFAULT_PORT_RANGE = [9222, 9230];
const PROBE_TIMEOUT_MS = 500;

export class DiscoveryError extends Error {}

// ---------------------------------------------------------------------------
// Cache file helpers
// ---------------------------------------------------------------------------

function cacheFilePath() {
  if (process.env.TV_MCP_CACHE_FILE) {
    return process.env.TV_MCP_CACHE_FILE;
  }
  const base = process.env.XDG_CACHE_HOME || path.join(os.homedir(), '.cache');
  return path.join(base, 'tv-mcp', 'cdp-port.json');
}

async function readCache() {
  const p = cacheFilePath();
  try {
    const raw = await fs.readFile(p, 'utf-8');
    const data = JSON.parse(raw);
    return { host: String(data.host), port: Number(data.port) };
  } catch {
    return null;
  }
}

async function writeCache(host, port) {
  const p = cacheFilePath();
  const dir = path.dirname(p);
  try {
    await fs.mkdir(dir, { recursive: true });
  } catch (err) {
    process.stderr.write(`[cdp_discover] cannot create cache dir ${dir}: ${err.message}\n`);
    return;
  }
  const payload = {
    host,
    port,
    discovered_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
    probe_url: `http://${host}:${port}/json/version`,
  };
  // Atomic write: write to a sibling tmp file, then rename.
  const tmp = path.join(dir, `.tmp_cdp-port-${process.pid}-${Date.now()}.json`);
  try {
    await fs.writeFile(tmp, JSON.stringify(payload), 'utf-8');
    await fs.rename(tmp, p);
  } catch (err) {
    process.stderr.write(`[cdp_discover] cache write failed: ${err.message}\n`);
    try { await fs.unlink(tmp); } catch {}
  }
}

// ---------------------------------------------------------------------------
// Port probing
// ---------------------------------------------------------------------------

function httpGetJson(url, timeoutMs = PROBE_TIMEOUT_MS) {
  return new Promise(resolve => {
    let settled = false;
    const finish = (value) => { if (!settled) { settled = true; resolve(value); } };

    const req = http.get(url, { timeout: timeoutMs }, res => {
      if (res.statusCode !== 200) {
        res.resume();
        return finish(null);
      }
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        try { finish(JSON.parse(Buffer.concat(chunks).toString('utf-8'))); }
        catch { finish(null); }
      });
      res.on('error', () => finish(null));
    });
    req.on('error', () => finish(null));
    req.on('timeout', () => { req.destroy(); finish(null); });
  });
}

async function isTvEndpoint(host, port) {
  const version = await httpGetJson(`http://${host}:${port}/json/version`);
  if (!version) return false;
  const targets = await httpGetJson(`http://${host}:${port}/json/list`);
  if (!Array.isArray(targets)) return false;
  return targets.some(t =>
    t && t.type === 'page' && /tradingview/i.test(t.url || '')
  );
}

function parsePortRange(spec) {
  if (!spec) return DEFAULT_PORT_RANGE;
  const m = String(spec).match(/^(\d+)-(\d+)$/);
  if (!m) {
    process.stderr.write(`[cdp_discover] invalid TV_MCP_PORT_RANGE=${spec} — using default\n`);
    return DEFAULT_PORT_RANGE;
  }
  const lo = Number(m[1]), hi = Number(m[2]);
  if (lo >= 1 && lo <= hi && hi <= 65535) return [lo, hi];
  process.stderr.write(`[cdp_discover] invalid TV_MCP_PORT_RANGE=${spec} — using default\n`);
  return DEFAULT_PORT_RANGE;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Resolve {host, port} for the TradingView CDP endpoint.
 * @param {{forceRefresh?: boolean}} opts forceRefresh=true skips the cache
 *   validation path (useful after a mid-session failure to force a rescan).
 *   Env CDP_PORT still wins.
 */
export async function discoverCdp(opts = {}) {
  const { forceRefresh = false } = opts;

  // 1. Explicit env override.
  if (process.env.CDP_PORT) {
    const port = Number(process.env.CDP_PORT);
    if (Number.isFinite(port) && port > 0) {
      return { host: process.env.CDP_HOST || DEFAULT_HOST, port };
    }
    process.stderr.write(`[cdp_discover] invalid CDP_PORT=${process.env.CDP_PORT} — falling back to discovery\n`);
  }

  const host = process.env.CDP_HOST || DEFAULT_HOST;

  // 2. Fast path: validate the cached port.
  if (!forceRefresh) {
    const cached = await readCache();
    if (cached) {
      const probeHost = host !== DEFAULT_HOST ? host : cached.host;
      if (await isTvEndpoint(probeHost, cached.port)) {
        process.stderr.write(`[cdp_discover] cache hit: ${probeHost}:${cached.port}\n`);
        return { host: probeHost, port: cached.port };
      }
      process.stderr.write(`[cdp_discover] cache miss at ${probeHost}:${cached.port} — scanning\n`);
    }
  }

  // 3. Scan range.
  const [lo, hi] = parsePortRange(process.env.TV_MCP_PORT_RANGE);
  const matches = [];
  for (let port = lo; port <= hi; port++) {
    if (await isTvEndpoint(host, port)) matches.push(port);
  }

  if (matches.length === 0) {
    throw new DiscoveryError(
      `No TradingView CDP endpoint found at ${host}:${lo}..${hi}. ` +
      'Ensure TradingView Desktop is running with --remote-debugging-port=<PORT> ' +
      'AND at least one chart tab is open. To bypass discovery, set CDP_PORT explicitly.'
    );
  }

  // Prefer cached port if alive; otherwise lowest match.
  const cached = !forceRefresh ? await readCache() : null;
  const chosen = (cached && matches.includes(cached.port)) ? cached.port : matches[0];

  if (matches.length > 1) {
    process.stderr.write(
      `[cdp_discover] ${matches.length} TV endpoints found at ${host} (ports=${matches.join(',')}); using ${chosen}\n`
    );
  } else {
    process.stderr.write(`[cdp_discover] scanned ${lo}-${hi}, found TV on ${host}:${chosen}\n`);
  }

  await writeCache(host, chosen);
  return { host, port: chosen };
}

/**
 * Synchronous helper for the rare case where a fully-resolved (host, port)
 * is already in the cache and the caller can't await. Returns null on miss.
 */
export function readCachedSync() {
  try {
    const raw = fssync.readFileSync(cacheFilePath(), 'utf-8');
    const data = JSON.parse(raw);
    return { host: String(data.host), port: Number(data.port) };
  } catch {
    return null;
  }
}

File: data_svc/cdp_discover.py
"""
Intelligent CDP port discovery for the TradingView Desktop integration.

Why: TradingView Desktop's --remote-debugging-port is not always 9222. Other
Chromium-based apps (Antigravity, Chrome, dev tools) may grab 9222 first,
pushing TV to 9223+. Hardcoding any single port makes the stack brittle.

Algorithm:
  1. If env CDP_PORT is set, use it verbatim (operator escape hatch).
  2. Probe the cached port (if a cache file exists). If the host:port still
     hosts a TV instance, return it (fast path — no scan).
  3. Scan the configured port range. The first port whose `/json/list`
     contains a tab matching tradingview.com/chart wins. Fallback: any
     tab URL containing "tradingview" (case-insensitive).
  4. Write the chosen endpoint to the cache file atomically.

The same algorithm and cache file format are implemented in
external/tradingview-mcp/src/cdp_discover.js so the Node MCP server and
the Python data-svc converge on the same port without coordination.

Cache file:
  Path:   ${TV_MCP_CACHE_FILE} || ${XDG_CACHE_HOME or ~/.cache}/tv-mcp/cdp-port.json
  Schema: {"host": str, "port": int, "discovered_at": ISO-8601, "probe_url": str}

Env overrides:
  CDP_PORT             explicit port (skip discovery)
  CDP_HOST             host (defaults to "localhost")
  TV_MCP_CACHE_FILE    cache file path override
  TV_MCP_PORT_RANGE    port range "low-high" (default "9222-9230")
"""

from __future__ import annotations

import json
import logging
import os
import re
import socket
import tempfile
import urllib.error
import urllib.request
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional, Tuple

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Defaults
# ---------------------------------------------------------------------------

_DEFAULT_HOST = "localhost"
_DEFAULT_PORT_RANGE = (9222, 9230)
_PROBE_TIMEOUT_S = 0.5  # short — we may probe several ports back-to-back

_TV_CHART_PATTERN = re.compile(r"tradingview\.com/chart", re.IGNORECASE)
_TV_GENERIC_PATTERN = re.compile(r"tradingview", re.IGNORECASE)


class DiscoveryError(RuntimeError):
    """Raised when no TradingView CDP endpoint can be found."""


# ---------------------------------------------------------------------------
# Cache file helpers
# ---------------------------------------------------------------------------

def _cache_file_path() -> Path:
    override = os.environ.get("TV_MCP_CACHE_FILE")
    if override:
        return Path(override).expanduser()
    base = os.environ.get("XDG_CACHE_HOME") or os.path.join(
        os.path.expanduser("~"), ".cache"
    )
    return Path(base) / "tv-mcp" / "cdp-port.json"


def _read_cache() -> Optional[Tuple[str, int]]:
    path = _cache_file_path()
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text())
        host = str(data["host"])
        port = int(data["port"])
        return host, port
    except (json.JSONDecodeError, KeyError, ValueError, OSError) as exc:
        logger.debug("[cdp_discover] cache read failed: %s", exc)
        return None


def _write_cache(host: str, port: int) -> None:
    path = _cache_file_path()
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
    except OSError as exc:
        logger.warning("[cdp_discover] cannot create cache dir %s: %s", path.parent, exc)
        return

    payload = {
        "host": host,
        "port": port,
        "discovered_at": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "probe_url": f"http://{host}:{port}/json/version",
    }
    # Atomic write: write to a sibling tmp file, then rename.
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", dir=str(path.parent), prefix=".tmp_", suffix=".json", delete=False
        ) as tf:
            json.dump(payload, tf)
            tmp_name = tf.name
        os.replace(tmp_name, str(path))
    except OSError as exc:
        logger.warning("[cdp_discover] cannot write cache file %s: %s", path, exc)


# ---------------------------------------------------------------------------
# Port probing
# ---------------------------------------------------------------------------

def _http_get_json(url: str, timeout: float = _PROBE_TIMEOUT_S) -> Optional[object]:
    try:
        with urllib.request.urlopen(url, timeout=timeout) as resp:
            return json.loads(resp.read().decode("utf-8", errors="replace"))
    except (urllib.error.URLError, socket.timeout, json.JSONDecodeError, OSError):
        return None


def _is_tv_endpoint(host: str, port: int) -> bool:
    """True iff `host:port` exposes CDP with at least one TradingView tab open."""
    if _http_get_json(f"http://{host}:{port}/json/version") is None:
        return False
    targets = _http_get_json(f"http://{host}:{port}/json/list")
    if not isinstance(targets, list):
        return False
    for t in targets:
        if not isinstance(t, dict) or t.get("type") != "page":
            continue
        url = t.get("url", "") or ""
        if _TV_GENERIC_PATTERN.search(url):
            return True
    return False


def _parse_port_range(spec: Optional[str]) -> Tuple[int, int]:
    if not spec:
        return _DEFAULT_PORT_RANGE
    try:
        lo_s, hi_s = spec.split("-", 1)
        lo, hi = int(lo_s), int(hi_s)
        if 1 <= lo <= hi <= 65535:
            return lo, hi
    except (ValueError, AttributeError):
        pass
    logger.warning("[cdp_discover] invalid TV_MCP_PORT_RANGE=%r — using default", spec)
    return _DEFAULT_PORT_RANGE


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def discover_cdp(force: bool = False) -> Tuple[str, int]:
    """
    Resolve (host, port) for the TradingView CDP endpoint.

    `force=True` skips the cache (useful after a probe failure mid-session
    to force a rescan; env CDP_PORT still wins).
    """
    # 1. Explicit env override.
    env_port = os.environ.get("CDP_PORT")
    if env_port:
        host = os.environ.get("CDP_HOST", _DEFAULT_HOST)
        try:
            return host, int(env_port)
        except ValueError:
            logger.warning("[cdp_discover] invalid CDP_PORT=%r — falling back to discovery", env_port)

    host = os.environ.get("CDP_HOST", _DEFAULT_HOST)

    # 2. Fast path: validate the cached port.
    if not force:
        cached = _read_cache()
        if cached:
            c_host, c_port = cached
            # Honor env CDP_HOST if it changed since last cache write.
            probe_host = host if host != _DEFAULT_HOST else c_host
            if _is_tv_endpoint(probe_host, c_port):
                logger.info("[cdp_discover] cache hit: %s:%d", probe_host, c_port)
                return probe_host, c_port
            logger.info("[cdp_discover] cache miss at %s:%d — scanning", probe_host, c_port)

    # 3. Scan range.
    lo, hi = _parse_port_range(os.environ.get("TV_MCP_PORT_RANGE"))
    matches: list[int] = []
    for port in range(lo, hi + 1):
        if _is_tv_endpoint(host, port):
            matches.append(port)

    if not matches:
        raise DiscoveryError(
            f"No TradingView CDP endpoint found at {host}:{lo}..{hi}. "
            "Ensure TradingView Desktop is running with --remote-debugging-port=<PORT> "
            "AND at least one chart tab is open. To bypass discovery, set CDP_PORT explicitly."
        )

    # Prefer the cached port if it's still among the matches (port stability
    # across reboots). Otherwise pick the lowest.
    cached = _read_cache() if not force else None
    chosen: int
    if cached and cached[1] in matches:
        chosen = cached[1]
    else:
        chosen = matches[0]

    if len(matches) > 1:
        logger.warning(
            "[cdp_discover] %d TradingView endpoints found at %s (ports=%s); using %d",
            len(matches), host, matches, chosen,
        )
    else:
        logger.info("[cdp_discover] scanned %d-%d, found TV on %s:%d", lo, hi, host, chosen)

    _write_cache(host, chosen)
    return host, chosen
